File: app/javascript/channels/chat_room_channel.js
document.addEventListener("DOMContentLoaded", () => {
  const chatRoomId = document.getElementById("message-form")?.dataset.chatRoomId; // Get chat_room ID

  if (!chatRoomId) {
    console.error("Chat Room ID is missing!");
    return;
  }

  console.log("Chat Room ID:", chatRoomId);
  console.log("Consumer:", window.consumer);

  if (!window.consumer) {
    console.error("ActionCable consumer is missing!");
    return;
  }

  const chatRoomChannel = window.consumer.subscriptions.create(
    { channel: "ChatRoomChannel", room_id: chatRoomId },
    {
      received(data) {
        console.log("New Message Received:", data.message);
        const messageElement = document.createElement("p");
        messageElement.innerHTML = `<strong>User:</strong> ${data.message}`;
        document.getElementById("messages").appendChild(messageElement);
      }
    }
  );

  document.getElementById("message-form").addEventListener("submit", function (e) {
    e.preventDefault();
    const message = document.getElementById("message-input").value;

    console.log("Sending message:", message);
    console.log("Consumer Object:", window.consumer);

    fetch(`/chat_rooms/${chatRoomId}/messages`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": document.querySelector('meta[name="csrf-token"]').content
      },
      body: JSON.stringify({ message: { content: message } })
    })
      .then(response => response.json())
      .then(data => console.log("Message sent successfully:", data))
      .catch(error => console.error("Error sending message:", error));

    document.getElementById("message-input").value = "";
  });
}); // ✅ Correct closing of the event listener

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
   def subscribed
  
    if params[:room_id].present?
      chat_room = ChatRoom.find_by(id: params[:room_id])
      if chat_room
        stream_from "chat_room_#{chat_room.id}"
        puts "✅ Subscribed to ChatRoom #{chat_room.id}"
      else
        reject_subscription
      end
    else
      puts "❌ ChatRoom ID is missing in params"
      reject_subscription
    end
  end

  def receive(data)
    chat_room = ChatRoom.find_by(id: data["room_id"])
    if chat_room
      message = chat_room.messages.create!(content: data["message"])
      ChatRoomChannel.broadcast_to(chat_room, { message: message.content })
      puts "✅ Message broadcasted: #{message.content}"
    else
      puts "❌ ChatRoom not found for received message"
    end
  end

end
